test/cases/migration_test.rb

```diff
@@ -101,76 +103,38 @@ def setup
       Person.connection.drop_table(:delete_me) rescue nil
     end
 
-    def test_adding_multiple_columns
-      classname = ActiveRecord::Base.connection.class.name[/[^:]*$/]
-      expected_query_count = {
-        "CockroachDBAdapter" => 8, # 5 new explicit columns, 2 for created_at/updated_at, 1 comment
-      }.fetch(classname) {
-          raise "need an expected query count for #{classname}"
-        }
-
-        assert_queries(expected_query_count) do
-          with_bulk_change_table do |t|
-            t.column :name, :string
-            t.string :qualification, :experience
-            t.integer :age, default: 0
-            t.date :birthdate, comment: "This is a comment"
-            t.timestamps null: true
-          end
-        end
-
-        assert_equal 8, columns.size
-        [:name, :qualification, :experience].each { |s| assert_equal :string, column(s).type }
-        assert_equal "0", column(:age).default
-        assert_equal "This is a comment", column(:birthdate).comment
-    end
-
-    def test_adding_indexes
-      with_bulk_change_table do |t|
-        t.string :username
-        t.string :name
-        t.integer :age
-      end
-
-      classname = ActiveRecord::Base.connection.class.name[/[^:]*$/]
-      expected_query_count = {
-        "CockroachDBAdapter" => 2,
-      }.fetch(classname) {
-          raise "need an expected query count for #{classname}"
-        }
-
-        assert_queries(expected_query_count) do
-          with_bulk_change_table do |t|
-            t.index :username, unique: true, name: :awesome_username_index
-            t.index [:name, :age]
-          end
+    %i(
+      test_adding_indexes
+      test_removing_index
+      test_adding_multiple_columns
+      test_changing_index
+    ).each do |method_name|
+      file, line = ::BulkAlterTableMigrationsTest.instance_method(method_name).source_location
+      iter = File.foreach(file)
+      (line - 1).times { iter.next }
+      indent = iter.next[/\A\s*/]
+      content = +""
+      content << iter.next until iter.peek == indent + "end\n"
+      content['"PostgreSQLAdapter" =>'] = '"CockroachDBAdapter" =>'
+      eval(<<-RUBY, binding, __FILE__, __LINE__ + 1)
+        def #{method_name}
+          #{content}
```

Comment:
I think we should go more in that direction to avoid having micro changes in the tests making our test suite fail. WDYT?

Reply:
i don't quite understand this change, could you explain more?

Reply:
Yeah definitely sorry. Basically I'm advocating for a bit of meta-programmation in tests.

## now

We currently are copying failing tests, and the related private methods. And then changing the bits we want to change.

## the change

I'd say that if the change is trivial, we can take advantage of ruby's meta-programmation tools to just change the parts we want. This would be way more resilient to little changes in the tests.

## the benefit and cost

This would make following rails updates so much easier, and even between minors or patches there are these kind of changes. It would reduce my workload. But it would make these local versions way harder to understand, as one would have to rely on understanding the meta change. But I think with good comment and viewing the original method, this should be ok

## the plan

I would make these changes progressively : with newly introduced tests, or with the one that are currently ignored and that we should stop ignoring anyway. Once there are a few examples I could refactor the meta-prog tools used (such as search and replace in the original function) to make the changes clearer and the footprint smaller

Reply:
thanks for explaining. i think the approach sounds good! i'd definitely like to make your maintenance burden easier.